File: frontend/src/api/notifications.ts
/**
 * 通知管理相关API
 */
import request from '@/utils/request'

// 通知类型枚举
export enum NotificationType {
  SYSTEM = 'system',          // 系统通知
  BOOKING = 'booking',        // 预约相关
  PAYMENT = 'payment',        // 支付相关
  COMPETITION = 'competition',  // 比赛相关
  EVALUATION = 'evaluation',  // 评价相关
  COACH_STUDENT = 'coach_student'  // 师生关系相关
}

// 通知优先级枚举
export enum NotificationPriority {
  LOW = 'low',        // 低
  NORMAL = 'normal',  // 普通
  HIGH = 'high',      // 高
  URGENT = 'urgent'   // 紧急
}

// 基础通知信息
export interface NotificationBase {
  title: string
  content: string
  type: string
  priority: string
  recipient_id: number
  resource_type?: string
  resource_id?: number
}

// 创建通知
export interface NotificationCreate extends NotificationBase {
  sender_id?: number
  send_email?: boolean
  send_sms?: boolean
  send_push?: boolean
  scheduled_at?: string
  expires_at?: string
}

// 更新通知
export interface NotificationUpdate {
  is_read?: boolean
  is_deleted?: boolean
}

// 通知响应
export interface NotificationResponse extends NotificationBase {
  id: number
  sender_id?: number
  is_read: boolean
  read_at?: string
  is_deleted: boolean
  send_email: boolean
  send_sms: boolean
  send_push: boolean
  scheduled_at?: string
  sent_at?: string
  expires_at?: string
  created_at: string
  updated_at?: string
}

// 通知查询参数
export interface NotificationQuery {
  type?: string
  is_read?: boolean
  priority?: string
  start_date?: string
  end_date?: string
  page?: number
  size?: number
}

// 通知模板
export interface NotificationTemplate {
  id: number
  code: string
  name: string
  type: string
  title_template: string
  content_template: string
  default_priority: string
  default_send_email: boolean
  default_send_sms: boolean
  default_send_push: boolean
  is_active: boolean
  created_at: string
  updated_at?: string
}

// 用户通知设置
export interface UserNotificationSettings {
  id: number
  user_id: number
  system_notifications: boolean
  booking_notifications: boolean
  payment_notifications: boolean
  competition_notifications: boolean
  evaluation_notifications: boolean
  coach_student_notifications: boolean
  email_enabled: boolean
  sms_enabled: boolean
  push_enabled: boolean
  quiet_start_time: string
  quiet_end_time: string
  weekend_quiet: boolean
  created_at: string
  updated_at?: string
}

// 更新用户通知设置
export interface UserNotificationSettingsUpdate {
  system_notifications?: boolean
  booking_notifications?: boolean
  payment_notifications?: boolean
  competition_notifications?: boolean
  evaluation_notifications?: boolean
  coach_student_notifications?: boolean
  email_enabled?: boolean
  sms_enabled?: boolean
  push_enabled?: boolean
  quiet_start_time?: string
  quiet_end_time?: string
  weekend_quiet?: boolean
}

// 通知统计
export interface NotificationStatistics {
  total_notifications: number
  unread_notifications: number
  read_notifications: number
  notifications_by_type: Array<{
    type: string
    count: number
  }>
  notifications_by_priority: Array<{
    priority: string
    count: number
  }>
}

// 批量通知
export interface BulkNotificationCreate {
  title: string
  content: string
  type: string
  priority: string
  recipient_ids: number[]
  send_email?: boolean
  send_sms?: boolean
  send_push?: boolean
  scheduled_at?: string
  expires_at?: string
}

export const notificationApi = {
  // 创建通知
  createNotification: (data: NotificationCreate) => {
    return request.post<NotificationResponse>('/notifications', data)
  },

  // 批量发送通知
  createBulkNotifications: (data: BulkNotificationCreate) => {
    return request.post<NotificationResponse[]>('/notifications/bulk', data)
  },

  // 获取通知列表
  getNotifications: (params?: NotificationQuery) => {
    return request.get<NotificationResponse[]>('/notifications', { params })
  },

  // 获取通知详情
  getNotification: (id: number) => {
    return request.get<NotificationResponse>(`/notifications/${id}`)
  },

  // 更新通知状态
  updateNotification: (id: number, data: NotificationUpdate) => {
    return request.put<NotificationResponse>(`/notifications/${id}`, data)
  },

  // 标记为已读
  markAsRead: (id: number) => {
    return request.post<NotificationResponse>(`/notifications/${id}/read`)
  },

  // 标记所有为已读
  markAllAsRead: () => {
    return request.post('/notifications/mark-all-read')
  },

  // 删除通知
  deleteNotification: (id: number) => {
    return request.delete(`/notifications/${id}`)
  },

  // 获取未读数量
  getUnreadCount: () => {
    return request.get<{ unread_count: number }>('/notifications/unread/count')
  },

  // 获取通知统计
  getStatistics: () => {
    return request.get<NotificationStatistics>('/notifications/statistics/summary')
  },

  // 获取我的通知设置
  getMySettings: () => {
    return request.get<UserNotificationSettings>('/notifications/settings/me')
  },

  // 更新我的通知设置
  updateMySettings: (data: UserNotificationSettingsUpdate) => {
    return request.put<UserNotificationSettings>('/notifications/settings/me', data)
  },

  // 获取通知模板列表
  getTemplates: () => {
    return request.get<NotificationTemplate[]>('/notifications/templates')
  },

  // 使用模板发送通知
  sendFromTemplate: (templateCode: string, recipientId: number, variables: Record<string, any> = {}) => {
    return request.post<NotificationResponse>(
      `/notifications/templates/${templateCode}/send?recipient_id=${recipientId}`,
      variables
    )
  }
}

export default notificationApi

File: backend/app/api/v1/notifications.py
"""
通知管理API
"""
from typing import List, Dict, Any
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session

from ...db.database import get_db
from ...models.user import User
from ...core.deps import get_current_user
from ...services.notification_service import NotificationService
from ...schemas.notification import (
    NotificationCreate, NotificationUpdate, NotificationResponse, NotificationQuery,
    NotificationTemplateCreate, NotificationTemplateUpdate, NotificationTemplateResponse,
    UserNotificationSettingsUpdate, UserNotificationSettingsResponse,
    NotificationStatistics, BulkNotificationCreate
)

router = APIRouter()


@router.post("/", response_model=NotificationResponse, summary="创建通知")
def create_notification(
    notification_data: NotificationCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    创建通知
    
    - 验证接收者存在性
    - 检查用户通知设置
    - 自动设置发送时间
    """
    notification = NotificationService.create_notification(db, notification_data, current_user)
    return NotificationResponse.from_orm(notification)


@router.post("/bulk", response_model=List[NotificationResponse], summary="批量发送通知")
def create_bulk_notifications(
    bulk_data: BulkNotificationCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    批量发送通知
    
    - 只有管理员可以批量发送
    - 同时发送给多个接收者
    - 自动处理每个用户的通知设置
    """
    notifications = NotificationService.create_bulk_notifications(db, bulk_data, current_user)
    return [NotificationResponse.from_orm(notif) for notif in notifications]


@router.get("/", response_model=List[NotificationResponse], summary="获取通知列表")
def get_notifications(
    type: str = Query(None, description="通知类型筛选"),
    is_read: bool = Query(None, description="是否已读筛选"),
    priority: str = Query(None, description="优先级筛选"),
    page: int = Query(1, ge=1, description="页码"),
    size: int = Query(20, ge=1, le=100, description="每页大小"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    获取用户通知列表
    
    - 支持按类型、状态、优先级筛选
    - 支持分页
    - 按创建时间倒序排列
    """
    query = NotificationQuery(
        type=type,
        is_read=is_read,
        priority=priority,
        page=page,
        size=size
    )
    notifications = NotificationService.get_notifications(db, current_user.id, query)
    return [NotificationResponse.from_orm(notif) for notif in notifications]


@router.get("/{notification_id}", response_model=NotificationResponse, summary="获取通知详情")
def get_notification(
    notification_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """获取通知详情"""
    notification = NotificationService.get_notification(db, notification_id, current_user.id)
    return NotificationResponse.from_orm(notification)


@router.put("/{notification_id}", response_model=NotificationResponse, summary="更新通知状态")
def update_notification(
    notification_id: int,
    notification_data: NotificationUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    更新通知状态
    
    - 标记已读/未读
    - 软删除通知
    - 自动设置阅读时间
    """
    notification = NotificationService.update_notification(db, notification_id, notification_data, current_user.id)
    return NotificationResponse.from_orm(notification)


@router.post("/{notification_id}/read", response_model=NotificationResponse, summary="标记为已读")
def mark_as_read(
    notification_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """标记通知为已读"""
    notification = NotificationService.mark_as_read(db, notification_id, current_user.id)
    return NotificationResponse.from_orm(notification)


@router.post("/mark-all-read", summary="标记所有通知为已读")
def mark_all_as_read(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    标记所有通知为已读
    
    - 批量更新未读通知
    - 返回更新数量
    """
    updated_count = NotificationService.mark_all_as_read(db, current_user.id)
    return {"message": f"已标记 {updated_count} 条通知为已读"}


@router.delete("/{notification_id}", summary="删除通知")
def delete_notification(
    notification_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """删除通知（软删除）"""
    NotificationService.delete_notification(db, notification_id, current_user.id)
    return {"message": "通知已删除"}


@router.get("/unread/count", summary="获取未读通知数量")
def get_unread_count(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    获取未读通知数量
    
    - 用于前端显示小红点
    - 实时更新
    """
    count = NotificationService.get_unread_count(db, current_user.id)
    return {"unread_count": count}


@router.get("/statistics/summary", response_model=NotificationStatistics, summary="获取通知统计")
def get_statistics(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    获取通知统计信息
    
    - 总数、已读、未读统计
    - 按类型和优先级分组统计
    """
    return NotificationService.get_statistics(db, current_user.id)


# 用户通知设置相关API
@router.get("/settings/me", response_model=UserNotificationSettingsResponse, summary="获取我的通知设置")
def get_my_settings(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """获取当前用户的通知设置"""
    settings = NotificationService.get_user_settings(db, current_user.id)
    return UserNotificationSettingsResponse.from_orm(settings)


@router.put("/settings/me", response_model=UserNotificationSettingsResponse, summary="更新我的通知设置")
def update_my_settings(
    settings_data: UserNotificationSettingsUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    更新当前用户的通知设置
    
    - 控制各类通知开关
    - 设置发送方式偏好
    - 配置免打扰时间
    """
    settings = NotificationService.update_user_settings(db, current_user.id, settings_data)
    return UserNotificationSettingsResponse.from_orm(settings)


# 通知模板管理API（管理员专用）
@router.post("/templates", response_model=NotificationTemplateResponse, summary="创建通知模板")
def create_template(
    template_data: NotificationTemplateCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    创建通知模板
    
    - 只有管理员可以创建
    - 支持变量占位符
    - 设置默认发送方式
    """
    template = NotificationService.create_template(db, template_data, current_user)
    return NotificationTemplateResponse.from_orm(template)


@router.get("/templates", response_model=List[NotificationTemplateResponse], summary="获取通知模板列表")
def get_templates(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """获取通知模板列表"""
    templates = NotificationService.get_templates(db)
    return [NotificationTemplateResponse.from_orm(template) for template in templates]


@router.post("/templates/{template_code}/send", response_model=NotificationResponse, summary="使用模板发送通知")
def send_from_template(
    template_code: str,
    recipient_id: int = Query(..., description="接收者ID"),
    variables: Dict[str, Any] = {},
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    使用模板发送通知
    
    - 自动渲染模板变量
    - 应用模板默认设置
    - 支持个性化变量替换
    """
    notification = NotificationService.send_from_template(db, template_code, recipient_id, variables, current_user)
    return NotificationResponse.from_orm(notification)
